### DISCUSSION TITLE
False Positive jakarta.persistence.NoResultException when using Hibernate with Spring Data JPA

### DISCUSSION THREAD

**Discussion by fbutter** [open-telemetry/opentelemetry-java-instrumentation#12863] [Q&A]
Given a Spring JpaRepository returning an Optional
When returning an empty result
Then an jakarta.persistence.NoResultException is reported

In a first investigation I found:

1. hibernate-core-6.5.3 throws an jakarta.persistence.NoResultException when no result is found 
```
org.hibernate.query.spi.AbstractSelectionQuery#getSingleResult

@Override
public R getSingleResult() {
	try {
		final List<R> list = list();
		if ( list.isEmpty() ) {
			throw new NoResultException(
					String.format( "No result found for query [%s]", getQueryString() )
			);
		}
		return uniqueElement( list );
	}
	catch ( HibernateException e ) {
		throw getSession().getExceptionConverter().convert( e, getQueryOptions().getLockOptions() );
	}
}
```
2. spring-data-jpa-3.3.5 catches this exception and finally converts it to an empty Optional
```
org.springframework.data.jpa.repository.query.JpaQueryExecution#execute
@Nullable
public Object execute(AbstractJpaQuery query, JpaParametersParameterAccessor accessor) {

	Assert.notNull(query, "AbstractJpaQuery must not be null");
	Assert.notNull(accessor, "JpaParametersParameterAccessor must not be null");

	Object result;

	try {
		result = doExecute(query, accessor);
	} catch (NoResultException e) {
		return null;
	}

	if (result == null) {
		return null;
	}

	JpaQueryMethod queryMethod = query.getQueryMethod();
	Class<?> requiredType = queryMethod.getReturnType();

	if (ClassUtils.isAssignable(requiredType, void.class) || ClassUtils.isAssignableValue(requiredType, result)) {
		return result;
	}

	return CONVERSION_SERVICE.canConvert(result.getClass(), requiredType) //
			? CONVERSION_SERVICE.convert(result, requiredType) //
			: result;
}
```

3. The OTEL hibernate 6 instrumentation reports the jakarta.persistence.NoResultException as an error, see [QueryInstrumentation.java](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/hibernate/hibernate-6.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/hibernate/v6_0/QueryInstrumentation.java)

Is there a way around that?

**Comment by steverao** [comment]
Thank you for your report, there is a PR https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/12879 that fixed it.

**Reply by fbutter** [reply]
Thank you very much
